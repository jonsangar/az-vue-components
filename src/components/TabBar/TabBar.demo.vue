<template lang="pug">
  .tabBarDemo
    Demo(title="Uso Básico")
      az-tabBar(v-model="active" :fixed="false")
        az-tabBar-item(icon="palm") Tab
        az-tabBar-item(icon="lap") Tab
        az-tabBar-item(icon="desktop") Tab
        az-tabBar-item(icon="large") Tab

    Demo(title="Seleccionar por nombre")
      az-tabBar(v-model="name" :fixed="false")
        az-tabBar-item(icon="palm" name="palm")
        az-tabBar-item(icon="lap" name="lap")
        az-tabBar-item(icon="desktop" name="desktop")
        az-tabBar-item(icon="large" name="large")

    Demo(title="Barra de navegación")
      az-tabBar(v-model="route" :fixed="false")
        az-tabBar-item(icon="palm" to="home") Home
        az-tabBar-item(icon="lap" to="shop") Shop
        az-tabBar-item(icon="desktop" to="faqs") Faqs
        az-tabBar-item(icon="large" to="contact") Contact

    Demo(title="Mostrar etiqueta")
      az-tabBar(v-model="badge" :fixed="false")
        az-tabBar-item(icon="palm") Tab
        az-tabBar-item(icon="lap" dot) Tab
        az-tabBar-item(icon="desktop" info="5") Tab
        az-tabBar-item(icon="large" info="10") Tab

    Demo(title="Mostrar etiqueta")
      az-tabBar(v-model="custom" :fixed="false" active-color="var(--color-brand)" inactive-color="var(--color-gray1)")
        az-tabBar-item(icon="palm") Tab
        az-tabBar-item(icon="lap" dot) Tab
        az-tabBar-item(icon="desktop" info="5") Tab
        az-tabBar-item(icon="large" info="10") Tab
</template>

<script>
export default {
  data() {
    return {
      active: 0,
      name: 'palm',
      route: 0,
      badge: 0,
      custom: 0,
    };
  },
};
</script>

<style lang="scss">
.tabBarDemo {
  .az-tabBar {
    margin-bottom: simple();
    background-color: color(gray7);
  }
}
</style>
