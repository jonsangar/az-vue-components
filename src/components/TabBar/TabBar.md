---
device: true
---

# TabBar

### Instalación

```javascript
import { TabBar } from 'didor';

Vue.use(TabBar);
```

---

### Uso

#### Uso básico

``` html
<az-tabBar v-model="active">
  <az-tabBar-item icon="palm">Tab</az-tabBar-item>
  <az-tabBar-item icon="lap">Tab</az-tabBar-item>
  <az-tabBar-item icon="desktop">Tab</az-tabBar-item>
  <az-tabBar-item icon="large">Tab</az-tabBar-item>
</az-tabBar>
```

``` js
export default {
  data() {
    return {
      active: 0
    }
  }
}
```

#### Seleccionar por nombre

``` html
<az-tabBar v-model="active">
  <az-tabBar-item icon="palm" name="palm"/>
  <az-tabBar-item icon="lap" name="lap"/>
  <az-tabBar-item icon="desktop" name="desktop"/>
  <az-tabBar-item icon="large" name="large"/>
</az-tabBar>
```

``` js
export default {
  data() {
    return {
      active: 0
    }
  }
}
```

#### Barra de navegación

Cuando funciona como barra de navegación, no es necesario indicar el item activo, lo toma directamente de la ruta de navegación.

Debemos pasar el nombre de la ruta o un objeto si queremos pasar parámetros:

``` html
<az-tabBar>
  <az-tabBar-item icon="palm" to="home">home</az-tabBar-item>
  <az-tabBar-item icon="lap" :to="{name: 'shop', params: {category: 'phones'}}">Shop</az-tabBar-item>
  <az-tabBar-item icon="desktop" to="faqs">Faqs</az-tabBar-item>
  <az-tabBar-item icon="large" to="contact">Contact</az-tabBar-item>
</az-tabBar>
```

#### Mostrar etiqueta

``` html
<az-tabBar v-model="active">
  <az-tabBar-item icon="palm">Tab</az-tabBar-item>
  <az-tabBar-item icon="lap" dot>Tab</az-tabBar-item>
  <az-tabBar-item icon="desktop" info="5">Tab</az-tabBar-item>
  <az-tabBar-item icon="large" info="10">Tab</az-tabBar-item>
</az-tabBar>
```

#### Colores personalizados

``` html
<az-tabBar v-model="active" active-color="var(--color-brand)" inactive-color="var(--color-gray1)">
  <az-tabBar-item icon="palm"/>
>Tab</az-tabBar-item>
  <az-tabBar-item icon="lap" dot>Tab</az-tabBar-item>
  <az-tabBar-item icon="desktop" info="5">Tab</az-tabBar-item>
  <az-tabBar-item icon="large" info="10">Tab</az-tabBar-item>
</az-tabBar>
```

---

### API TabBar

| Attribute   | Description                                | Type      | Default   |
| ----------- | ------------------------------------------ | --------- | --------- |
| v-model | Identifica la pestaña activa | `String` `Number` | `0` |
| fixed | Mantiene la barra fija en la parte de abajo de la pantalla | `Boolean` | `true` |
| safeArea | Habilita un espacio extra en la parte inferior de la barra para dispositivos con un botón en la parte inferior de la pantalla | `Boolean` | `true` |
| z-index | Permite especificar el valor del z-index de la barra | `Number` | `1` |
| active-Color | Color de un item activo | `String` | -- |
| inactive-Color | Color de un item inactivo | `String` | -- |

### Eventos TabBar

| Evento | Descripción                                             | Argumento |
| ------ | ------------------------------------------------------- | --------- |
| change | Se lanza cuando cambia el valor del TabBar | **active**: Index de la pestaña activa         |


### API TabBarItem

| Attribute   | Description                                | Type      | Default   |
| ----------- | ------------------------------------------ | --------- | --------- |
| icon | Nombre del icono | `String` | -- |
| name | Identificador de la pestaña | `String` `Number` | -- |
| dot | Muestra un punto rojo sobre el icono | `Boolean` | `false` |
| info | Muestra un mensaje en rojo sobre el icono | `String` | -- |
| to | Nombre de la ruta utilizando vue-router | `String` `Object` | -- |

### Eventos TabBarItem

| Evento | Descripción                                             | Argumento |
| ------ | ------------------------------------------------------- | --------- |
| click | Se lanza cuando hacemos click en un item del TabBar | -         |
