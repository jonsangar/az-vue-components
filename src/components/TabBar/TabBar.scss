$tabBar-height: double();
$tabBar-background-color: #ffffff;

.az-tabBar {
  $rootClass: &;

  display: flex;
  width: 100%;
  height: $tabBar-height;
  background-color: $tabBar-background-color;

  &--fixed {
    position: fixed;
    bottom: 0;
    left: 0;
  }

  &--safe-area {
    padding-bottom: constant(safe-area-inset-bottom);
    padding-bottom: env(safe-area-inset-bottom);
    padding-bottom: var(--notch-inset-bottom);
  }
}
