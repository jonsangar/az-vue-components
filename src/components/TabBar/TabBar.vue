<template lang="pug">
  .az-tabBar(:class="tabBarClass" :style="tabBarStyle" @keyup="onKeyup")
    slot
</template>

<script>
export default {
  props: {
    value: {
      type: [Number, String],
      default: 0,
    },
    activeColor: {
      type: String,
      required: false,
    },
    inactiveColor: {
      type: String,
      required: false,
    },
    fixed: {
      type: Boolean,
      default: true,
    },
    zIndex: {
      type: Number,
      default: 1,
    },
    safeArea: {
      type: Boolean,
      default: true,
    },
  },

  watch: {
    value: 'setActiveItem',
  },

  computed: {
    tabBarClass() {
      return [this.fixed ? 'az-tabBar--fixed' : '', this.safeArea ? 'az-tabBar--safeArea' : ''];
    },

    tabBarStyle() {
      return {
        'z-index': this.zIndex,
      };
    },
  },

  methods: {
    setActiveItem() {
      this.$children.forEach((item, index) => {
        if (item.to) return;
        item.active = (item.name || index) === this.value;
      });
    },

    onKeyup(e) {
      this.$emit('keyup',e);
    },

    onChange(active) {
      if (active !== this.value) {
        this.$emit('input', active);
        this.$emit('change', active);
      }
    },
  },

  mounted() {
    this.setActiveItem();
  },
};
</script>

<style src="./TabBar.scss" lang="scss"></style>
