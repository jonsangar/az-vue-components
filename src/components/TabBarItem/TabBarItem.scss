$tabBar-item-text-color: color(gray4);
$tabBar-item-active-color: color(gray1);
$tabBar-item-inactive-color: color(gray5);

.az-tabBarItem {
  $rootClass: &;

  cursor: pointer;
  position: relative;
  display: flex;
  flex: 0 0 auto;
  flex-basis: 0;
  flex-direction: column;
  flex-grow: 1;
  align-items: center;
  justify-content: center;
  height: 100%;
  padding: 0.2rem 0;
  font-size: font-size(small2);
  color: $tabBar-item-text-color;

  &__icon {
    position: relative;
    display: block;
    margin: 0;

    &--dot {
      &::after {
        position: absolute;
        top: 0.2rem;
        right: 0;
        width: 0.5rem;
        height: 0.5rem;
        content: ' ';
        background-color: color(danger);
        border-radius: 100%;
      }
    }
  }

  &__disabled { // lock icon
    position: absolute;
    bottom: -.1rem;
    right: -.1rem;
    height: fit-content;
    line-height: 0;
    color: color(gray4)
  }

  &__badge {
    position: absolute;
    top: 0;
    right: 0.2rem;
    min-width: 1rem;
    padding: 0 0.2rem;
    font-size: font-size(small2);
    font-weight: font-weight(bold);
    line-height: 0.9rem;
    color: #ffffff;
    text-align: center;
    background-color: color(danger);
    border-radius: 0.4rem;
    transform: translateX(50%);
    transform-origin: 100%;
  }

  &__title {
    position: absolute;
    bottom: 0;
    padding-bottom: 0.1rem;
    line-height: 1rem;
    width: 100%;
    text-align: center;
  }

  &--active {
    color: $tabBar-item-active-color;
    // color: $tabBar-item-text-color;
  }

  &--selected {
    color: color(brand);
  }

  &--title {
    padding-bottom: 1rem;
  }

  &--disabled {
    color: $tabBar-item-inactive-color;
  }
}

@include media(landscape){
  .az-tabBar{
    right: 0;
    left: 0;
    max-width: $palm;
    margin: 0 auto;

  }
}
