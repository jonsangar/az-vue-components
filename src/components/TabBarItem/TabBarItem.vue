<template lang="pug">
  .az-tabBarItem(
    :class="tabBarItemClass"
    :style="tabBarItemStyle"
    @click="onClick"
    @keyup="onKeyup")

    .az-tabBarItem__icon(
      v-if="icon"
      :class="{'az-tabBarItem__icon--dot': dot && !info}")

      //- Icono
      az-icon(:name="icon" :scale="scale")

      //- .az-tabBarItem__disabled(v-if="disabled")
      //-   az-icon(name="lock" scale=".5")

      //- Badge / Dot
      .az-tabBarItem__badge(v-if="info") {{ info }}

    //- title
    .az-tabBarItem__title(v-if="$slots.default")
      slot
</template>

<script>
export default {
  props: {
    icon: {
      type: String,
      required: false,
    },
    name: {
      type: [Number, String],
      required: false,
    },
    dot: {
      type: Boolean,
      default: false,
    },
    info: {
      type: String,
      required: false,
    },
    disabled: {
      type: Boolean,
      dafault: false
    },
    scale: {
      type: String,
      default: "1"
    },
    to: {
      type: [String, Object],
      required: false,
    },
  },

  data() {
    return {
      active: false,
    };
  },

  computed: {
    routeActive() {
      if (!this.to || !this.$router || this.$router.name) return false;

      const toName = typeof this.to === 'object' ? this.to.name : this.to;
      const routeName = this.$router.history.current.name;

      return toName === routeName;
    },

    index() {
      return this.$parent.$children.indexOf(this);
    },

    tabBarItemClass() {
      return [
        this.$slots.default ? 'az-tabBarItem--title' : '',
        this.active ? 'az-tabBarItem--active' : '',
        this.disabled ? 'az-tabBarItem--disabled' : ''
      ];
    },

    tabBarItemStyle() {
      return {
        color:
          this.active || this.routeActive
            ? this.$parent.$options.propsData.activeColor
            : this.$parent.$options.propsData.inactiveColor,
      };
    },
  },

  methods: {
    onClick(event) {
      this.$parent.onChange(this.name || this.index);
      this.$emit('click', event);

      if (this.to) {
        const route = typeof this.to === 'object' ? this.to : { name: this.to };
        this.$router.push(route).catch(() => {});
      }
    },

    onKeyup (event) {
      this.$emit('keyup', event);
    }
  },
};
</script>

<style src="./TabBarItem.scss" lang="scss"></style>
