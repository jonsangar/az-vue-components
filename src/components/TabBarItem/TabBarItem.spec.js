import { shallowMount, createLocalVue } from '@vue/test-utils';
// Importamos librerÃ­as de estilos e iconos
import SvgIcon from 'vue-svgicon'
import VueRouter from 'vue-router'
import AirtoolsBLERoutes from 'Fixtures/AirtoolsBLE.routes';
import TabBar from 'Core/components/TabBar/TabBar.vue';
import TabBarItem from './TabBarItem.vue';

const localVue = createLocalVue()
localVue.use(VueRouter)
localVue.use(SvgIcon, {
  tagName: 'az-icon',
  classPrefix: 'az',
})

const router = new VueRouter({
  routes: AirtoolsBLERoutes
})

describe('TabBarItem.vue', () => {

  const build = () => {
    const wrapper = shallowMount(TabBarItem, {
      localVue,
      parentComponent: TabBar,
      router,
      propsData: {
        value: true,
        icon: 'save',
        name: 'brand',
        dot: true,
        info: 'MyInfo',
        to: { name: 'airtools' },
        active: true,
      },
      computed: {
        tabBarItemStyle: () => jest.fn()
      },
      mocks: {
        $parent: {
          onChange: () => jest.fn()
        }
      }
    })

    return {
      wrapper,
    }
  }

  it('Comprueba renderizado del componente TabBarItem con snapshot', () => {
    const { wrapper } = build()
    expect(wrapper.html()).toMatchSnapshot()
  })

  it('Comprueba que se actualizan las props cuando se asignan las props', async () => {
    const { wrapper } = build()

    expect(wrapper.vm.icon).toBe('save')
    expect(wrapper.vm.name).toBe('brand')
    expect(wrapper.vm.dot).toBe(true)
    expect(wrapper.vm.info).toBe('MyInfo')
    expect(wrapper.vm.to).toEqual({ name: 'airtools' })
  });

  it('Comprueba que emite el evento change cuando se clica', async () => {
    const { wrapper } = build()

    const e = {
      event: {
        type: 'change'
      }
    }
    await wrapper.vm.onKeyup(e)
    expect(wrapper.emitted().keyup).toBeTruthy()
    expect(wrapper.emitted().keyup.length).toBe(1)
    expect(wrapper.emitted().keyup[0]).toEqual([e])
  });

  it('Comprueba que emite el evento click cuando se clica', async () => {
    const { wrapper } = build()
    await wrapper.vm.$nextTick()

    const e = {
      event: {
        type: 'click'
      }
    }

    await wrapper.vm.onClick(e)
    expect(wrapper.emitted().click).toBeTruthy()
    expect(wrapper.emitted().click.length).toBe(1)
    expect(wrapper.emitted().click[0]).toEqual([e])
  });
})
