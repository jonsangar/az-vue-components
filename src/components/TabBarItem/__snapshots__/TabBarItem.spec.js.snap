// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TabBarItem.vue Comprueba renderizado del componente TabBarItem con snapshot 1`] = `
<div class="az-tabBarItem" value="true" active="true">
  <div class="az-tabBarItem__icon">
    <az-icon-stub name="save" width="" height="" scale="1" fill="true"></az-icon-stub>
    <div class="az-tabBarItem__badge">MyInfo</div>
  </div>
  <!---->
</div>
`;
